django/views/decorators/http.py

```diff
@@ -40,25 +42,46 @@ def inner(request, *args, **kwargs):
                     request=request,
                 )
                 return response
-            return func(request, *args, **kwargs)
 
-        return inner
+        @wraps(func)
+        def _wrapper_view_sync(request, *args, **kwargs):
+            response = _process_request(request)
+            if response is not None:
+                return response
+
+            return func(request)
+
+        @wraps(func)
+        async def _wrapper_view_async(request, *args, **kwargs):
+            response = _process_request(request)
+            if response is not None:
+                return response
+
+            return await func(request)
+
+        if asyncio.iscoroutinefunction(func):
+            return _wrapper_view_async
+        return _wrapper_view_sync
```

Comment:
Again, when we're not actually consuming the wrapped view (when there's no process response) we should just be able to mark the wrapper as a coroutine, rather than duplicating it throughout.